plugins/fileservices/fileservices.cpp

```diff
@@ -198,63 +198,141 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
     return getHPCCEnvironment();
 }
 
-static StringAttr espurl;           // Default ESP url if none specified
+static void setServerAccess(CClientFileSpray &server, IConstWorkUnit * wu)
+{
+    StringBuffer user, password, token;
+    wu->getSecurityToken(StringBufferAdaptor(token));
+    extractToken(token.str(), wu->queryWuid(), StringBufferAdaptor(user), StringBufferAdaptor(password));
+    server.setUsernameToken(user.str(), password.str(), "");
+}
+
+typedef MapStringTo<time_t> StrMapToEpoch;
+static StrMapToEpoch espUrlsLastHeartBeat;
 static CriticalSection espURLcrit;
 
+static void setWsFSUrlAliveTimeStamp(const char * url, time_t aliveTime)
+{
+    StringBuffer key;
+    CriticalBlock b(espURLcrit);
+    espUrlsLastHeartBeat.setValue(key.set(url).str(), aliveTime);
+}
+
+static time_t contactWsFS(const char * espurl)
+{
+    CClientFileSpray server;
+    server.addServiceUrl(espurl);
+
+    try
+    {
+        Owned<IClientEchoDateTime> req = server.createEchoDateTimeRequest();
```

Comment:
How long does this take to time out if the server is NOT available? And is that a problem?